File: burstLibrary/src/main/java/com/harry1453/burst/explorer/entity/BurstAddress.java
package com.harry1453.burst.explorer.entity;

import com.harry1453.burst.BurstUtils;

import java.math.BigInteger;
import java.util.Objects;

public class BurstAddress {

    private final String address;
    private final BigInteger numericID;

    public BurstAddress(BigInteger numericID) {
        if (numericID == null) {
            this.numericID = BigInteger.ZERO;
        } else {
            this.numericID = numericID;
        }

        if (this.numericID.equals(BigInteger.ZERO)) {
            this.address = "";
        } else {
            this.address = BurstUtils.toBurstAddress(this.numericID);
        }
    }

    public BurstAddress(String RS) throws BurstUtils.ReedSolomon.DecodeException {
        if (RS.startsWith("BURST-")) {
            RS = RS.substring(6);
        }
        numericID = BurstUtils.toNumericID(RS);
        address = RS;
    }

    public BigInteger getNumericID() {
        return numericID;
    }

    public String getRawAddress() {
        return address;
    }

    public String getFullAddress() {
        if (address == null || address.length() == 0) {
            return "";
        } else {
            return "BURST-" + address;
        }
    }

    @Override
    public String toString() {
        return getFullAddress();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof BurstAddress && Objects.equals(numericID, ((BurstAddress) obj).numericID);
    }
}

File: burstLibrary/src/main/java/com/harry1453/burst/explorer/entity/BurstAddress.kt
package com.harry1453.burst.explorer.entity

import com.harry1453.burst.BurstUtils
import java.math.BigInteger

class BurstAddress {

    private var address: String
    private var numericID: BigInteger

    constructor(numericID: BigInteger?) {
        this.numericID = numericID ?: BigInteger.ZERO

        this.address = if (this.numericID == BigInteger.ZERO) {
            ""
        } else {
            BurstUtils.toBurstAddress(this.numericID)
        }
    }

    @Throws(BurstUtils.ReedSolomon.DecodeException::class)
    constructor(reedSolomonAddressIn: String?) {
        var reedSolomonAddress = reedSolomonAddressIn ?: ""
        if (reedSolomonAddress.startsWith("BURST-")) {
            reedSolomonAddress = reedSolomonAddress.substring(6)
        }
        numericID = BurstUtils.toNumericID(reedSolomonAddress)
        address = reedSolomonAddress
    }

    fun getNumericID(): BigInteger {
        return numericID
    }

    fun getRawAddress(): String {
        return address
    }

    fun getFullAddress(): String {
        return if (address.isEmpty()) {
            ""
        } else {
            "BURST-$address"
        }
    }

    override fun toString(): String {
        return getFullAddress()
    }

    override fun equals(other: Any?): Boolean {
        return other is BurstAddress && numericID == other.numericID
    }

    override fun hashCode(): Int {
        var result = address.hashCode()
        result = 31 * result + numericID.hashCode()
        return result
    }
}
